src/thermo/ConstDensityTabulatedThermo.cpp

```diff
@@ -0,0 +1,203 @@
+/**
+ *  @file ConstDensityTabulatedThermo.cpp
+ * Definition file for a derived class of ThermoPhase that assumes either
+ * an ideal gas or ideal solution approximation and handles
+ * variable pressure standard state methods for calculating
+ * thermodynamic properties (see \ref thermoprops and
+ * class \link Cantera::ConstDensityTabulatedThermo ConstDensityTabulatedThermo\endlink).
+ */
+/*
+ * Copyright (2005) Sandia Corporation. Under the terms of
+ * Contract DE-AC04-94AL85000 with Sandia Corporation, the
+ * U.S. Government retains certain rights in this software.
+ */
+
+#include "cantera/thermo/ConstDensityTabulatedThermo.h"
+#include "cantera/base/ctml.h"
+
+namespace Cantera
+{
+
+void ConstDensityTabulatedThermo::setMoleFractions(const doublereal* const x)
+{
+	Phase::setMoleFractions(x);
+	_updateThermo();
+}
+
+void ConstDensityTabulatedThermo::setMoleFractions_NoNorm(const doublereal* const x)
+{
+	Phase::setMoleFractions_NoNorm(x);
+	_updateThermo();
+}
+
+void ConstDensityTabulatedThermo::setMassFractions(const doublereal* const y)
+{
+	Phase::setMassFractions(y);
+	_updateThermo();
+}
+
+void ConstDensityTabulatedThermo::setMassFractions_NoNorm(const doublereal* const y)
+{
+	Phase::setMassFractions_NoNorm(y);
+	_updateThermo();
+}
+
+void ConstDensityTabulatedThermo::setConcentrations(const doublereal* const conc)
+{
+	Phase::setConcentrations(conc);
+	_updateThermo();
+}
+
+void ConstDensityTabulatedThermo::_updateThermo()
+{
+	doublereal tnow = temperature();
+	doublereal xnow = moleFraction(m_kk_mod);
+	doublereal c[4];
+	doublereal dS_corr = 0.0;
+	doublereal tlow = 0.0, thigh = 0.0;
+	int type = 0;
+	if (m_tlast != tnow || m_xlast != xnow) {
+		c[0]=tnow;
+		c[1]=interp_h(xnow) * 1e3; // 1e3 for conversion J/mol -> J/kmol
+		if (xnow==0) {
+			dS_corr = -BigNumber;
+		} else if (xnow==1) {
+			dS_corr = BigNumber;
+		} else {
+			dS_corr = GasConstant*std::log(xnow/(1.0-xnow));
+		}

```

Comment:
Why does the entropy of the "modifiable species" go to infinity as X->1? Shouldn't it just go to the pure species value, which should be well-defined?

Reply:
The `dS_corr` variable implements a correction term on the standard state entropy (and enthalpy), such that the calculated entropy of reaction matches the experimental value.  

<a href="https://www.codecogs.com/eqnedit.php?latex=\Delta&space;S^{\rm&space;expt}_{\rm&space;rxn}&space;=&space;\sum&space;\nu_i&space;s^{\rm&space;o}_i&space;&plus;&space;RT\ln\left(\prod_k&space;X_k^{\nu_k}&space;\right)" target="_blank"><img src="https://latex.codecogs.com/svg.latex?\Delta&space;S^{\rm&space;expt}_{\rm&space;rxn}&space;=&space;\sum&space;\nu_i&space;s^{\rm&space;o}_i&space;&plus;&space;RT\ln\left(\prod_k&space;X_k^{\nu_k}&space;\right)" title="\Delta S^{\rm expt}_{\rm rxn} = \sum \nu_i s^{\rm o}_i + RT\ln\left(\prod_k X_k^{\nu_k} \right)" /></a>

As X->1 the log term would go to +/- infinity, but from the tabulated data there is an actual value, so the standard entropy contains a term to correct for this log term.  Admittedly, there might be a more efficient way to do this, such that one is not adding the correction on, only to subtract it off, later.  I will also note that tabulated data (i.e. measured from experiments) typically does not approach X=1 or X=0, but typically stops well short of these limits.